### PR TITLE
Cleanups to record stack logic, prepare for adding factory config

### PR DESCRIPTION
**Pull request by jcredding**
This cleans up some of the record stack logic as preparation for
adding factory configs and using them in the record stack. Factory
configs will be used to allow storing factories that should be
used for an association and record stacks will need to use the
configured association factories as they walk associations. This
renames some classes and methods and updates the logic to match
our latest conventions.

This renames the stack `Record` and `Association` classes to
`RecordData` and `AssociationData`. These classes wrap a record
or association with stack specific behavior or helpers but calling
them a record or an association caused lots of confusion when
rereading the code. It's hard to keep straight what is an
activerecord record or association vs a stack record or
association. I previously used `stack_record` and
`stack_association` but even these are confusing because its not
a record or association that belongs to a stack. Generically
calling it data makes it clear its like a record or association
but also provides distinct naming between the two.

This also adds "record" to some methods and switches from a
"children" naming to just "associated". Adding record makes it
explicit about who the action is being done, i.e. you aren't
creating or destroying a tree node, its record is being created
or destroyed. Using "children" was also confusing in that the
associations are belongs to associations which are normally
considered a "parent" of the record not a "child".

Finally, this changes the `TreeNode`, `RecordData` and
`AssociationData` to be nested directly under `RecordStack`. This
fixes some issues in the tests where the full class name had to be
used and better matches how we typically nest objects in our
files.

@kellyredding - FYI. I'll merge these into the feature branch and we can discuss later. I pulled these changes out while working on the factory changes we discussed because I didn't want to clutter up those changes. This also gave me a good refresher on how the stack logic works.

### CODE REVIEW

File: test/unit/factory/record_stack_tests.rb [hunk 1]
```diff
@@ -239,176 +239,183 @@ class TreeNodeWithPresetAssociationsTests < UnitTests
...
+    should "set an association from a record data using `set_association`" do
+      another_record      = AnotherFakeRecord.new.tap(&:save!)
+      another_record_data = @record_data_class.new(another_record)
+
+      subject.set_association(:another, another_record_data)
       assert_equal another_record, @record.another
     end
 
-    should "reset associations, properly setting foreign keys, " \
+    should "reset associations to properly setting foreign keys " \
```
Comment (kellyredding) [inline test/unit/factory/record_stack_tests.rb]: @jcredding "reset associations to properly*set* foreign keys" (also trailing space).

File: test/unit/factory/record_stack_tests.rb [hunk 2]
```diff
@@ -239,176 +239,183 @@ class TreeNodeWithPresetAssociationsTests < UnitTests
...
-      assert_equal @association.klass,           subject.record_class
-      assert_equal @association.klass.to_s,      subject.key
-      assert_equal @record.another,              subject.preset_record
+      assert_equal @reflection.name,   subject.name
+      assert_equal @association.klass, subject.record_class
+      assert_equal @record.another,    subject.preset_record
     end
 
-    should "know if it was preset or not" do
+    should "know if its preset" do
```
Comment (kellyredding) [inline test/unit/factory/record_stack_tests.rb]: @jcredding "know if *it's* preset" (it is).

File: test/unit/factory/record_stack_tests.rb [hunk 3]
```diff
@@ -239,176 +239,183 @@ class TreeNodeWithPresetAssociationsTests < UnitTests
...
 
       @record.another = nil
-      sa = @stack_association_class.new(@record, @association)
-      assert_false sa.preset?
+      data = @association_data_class.new(@record, @reflection)
+      assert_false data.preset?
     end
 
-    should "be required if it is preset" do
+    should "know if its required" do
```
Comment (kellyredding) [inline test/unit/factory/record_stack_tests.rb]: @jcredding "know if *it's* required" (it is).

File: test/unit/factory/record_stack_tests.rb (hunk 1 above)
Comment (kellyredding) [inline test/unit/factory/record_stack_tests.rb]: @jcredding also there are a bunch of trailing spaces in the test descriptions - maybe do a sweep and add a blurb to the commit msg saying you cleaned this up bc you noticed it when doing these changes?  I'm not sure how many there are - probably only need to add a comment if there are a bunch more than just what I see in this diff.

Comment (kellyredding) [inline test/unit/factory/record_stack_tests.rb]: @jcredding oh yeah, you've already merged.  Apply these notes to any hotfix cleanup you do.

**Comment by jcredding** [comment]
@kellyredding - I merged this but could've waited. I thought I was gonna put the next one up but then I was like oh yeah let me try running it in our app and I found a bug. Thanks for looking at it, I'll make these changes in a hotfix or something.

File: test/unit/factory/record_stack_tests.rb (hunk 1 above)
Comment (jcredding) [inline test/unit/factory/record_stack_tests.rb]: @kellyredding - I think these are all line continuations, which is kind of an issue but they generate correct messages. I did find one test that had an extra space at the end but it was in like the fake query tests.

Comment (kellyredding) [inline test/unit/factory/record_stack_tests.rb]: @jcredding gah - didn't notice the line continuation.  thanks for clarifying.

**Pull request merged**